models/svc/vits/vits.py

```diff
@@ -0,0 +1,295 @@
+# adapted from https://github.com/svc-develop-team/so-vits-svc under AGPL-3.0 license
+import copy
+import torch
+from torch import nn
+from torch.nn import functional as F
+
+from utils.util import *
+from utils.f0 import f0_to_coarse
+
+from modules.transformer.attentions import Encoder
+from models.tts.vits.vits import ResidualCouplingBlock, PosteriorEncoder
+from models.vocoders.gan.generator.bigvgan import BigVGAN
+from models.vocoders.gan.generator.hifigan import HiFiGAN
+from models.vocoders.gan.generator.nsfhifigan import NSFHiFiGAN
+from models.vocoders.gan.generator.melgan import MelGAN
+from models.vocoders.gan.generator.apnet import APNet
+from modules.encoder.condition_encoder import ConditionEncoder
+
+
+
+def slice_pitch_segments(x, ids_str, segment_size=4):
+  ret = torch.zeros_like(x[:, :segment_size])
+  for i in range(x.size(0)):
+    idx_str = ids_str[i]
+    idx_end = idx_str + segment_size
+    ret[i] = x[i, idx_str:idx_end]
+  return ret
+
+def rand_slice_segments_with_pitch(x, pitch, x_lengths=None, segment_size=4):
+  b, d, t = x.size()
+  if x_lengths is None:
+    x_lengths = t
+  ids_str_max = x_lengths - segment_size + 1
+  ids_str = (torch.rand([b]).to(device=x.device) * ids_str_max).to(dtype=torch.long)
+  ret = slice_segments(x, ids_str, segment_size)
+  ret_pitch = slice_pitch_segments(pitch, ids_str, segment_size)
+  return ret, ret_pitch, ids_str
+
+class ContentEncoder(nn.Module):
+    def __init__(self,
+                 out_channels,
+                 hidden_channels,
+                 kernel_size,
+                 n_layers,
+                 gin_channels=0,
+                 filter_channels=None,
+                 n_heads=None,
+                 p_dropout=None):
+        super().__init__()
+        self.out_channels = out_channels
+        self.hidden_channels = hidden_channels
+        self.kernel_size = kernel_size
+        self.n_layers = n_layers
+        self.gin_channels = gin_channels
+        
+        self.f0_emb = nn.Embedding(256, hidden_channels)
+
+        self.enc_ = Encoder(
+            hidden_channels,
+            filter_channels,
+            n_heads,
+            n_layers,
+            kernel_size,
+            p_dropout)
+        
+        self.proj = nn.Conv1d(hidden_channels, out_channels * 2, 1)
+
+    # condition_encoder ver.
+    def forward(self, x, x_mask, noice_scale=1):
+        x = self.enc_(x * x_mask, x_mask)
+    # def forward(self, x, x_mask, f0=None, noice_scale=1):
+    #     x = x + self.f0_emb(f0).transpose(1, 2)
+    #     x = self.enc_(x * x_mask, x_mask)
+        stats = self.proj(x) * x_mask
+        m, logs = torch.split(stats, self.out_channels, dim=1)
+        z = (m + torch.randn_like(m) * torch.exp(logs) * noice_scale) * x_mask
+
+        return z, m, logs, x_mask
+
+class SynthesizerTrn(nn.Module):
```

Comment:
Is this class same as https://github.com/open-mmlab/Amphion/blob/42ca016cbafc9737e63cc1cb2ecfa2582cbcadf3/models/tts/vits/vits.py#L151

We need to merge them into a common one.

Reply:
No, the internal encoding is different, one for text and acoustic condition the other, which affects forward/infer functions either. So in my opinion it can not be merged.